### PR TITLE
dra: PodScheduling -> PodSchedulingHints

### LINKED ISSUE #114283: dynamic resource allocation: rename PodScheduling
The name is not ideal. This already came up during the initial API review and again in https://github.com/kubernetes/enhancements/issues/3063#issuecomment-1336225191:

@sftim wrote:
> Pretty much everything that has a kind in Kubernetes looks broadly idiomatic if you put a definite or indefinite article in front of it. However, PodScheduling is a bit weird.
> 
> (Fictional) examples:
> 
>  -   “the driver watches for a PodScheduling that matches this driver's resource type”
>  -    “you can use the kubectl tool to manually delete the PodScheduling”
>  -    “the PodScheduling has an ownerReference to the Pod and is therefore a dependent object of that Pod”

My own proposal was `PodSchedulingState`.

@sftim had some more:

- PodSchedulingHint
- PodSchedulingReservation
- ProvisionalPodBinding

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind cleanup
/kind api-change

#### What this PR does / why we need it:

The name "PodScheduling" was unusual because in contrast to most other names,
it was impossible to put an article in front of it.
    
Now PodSchedulingHints is used instead. It's always in plural form because a
single object represents a collection of hints. This is similar to "Endpoints".
    
#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes/kubernetes/issues/114283

#### Special notes for your reviewer:

To make this work technically, the client-go plural form exception list must be
extended. The hard-coded defaults cannot be changed (might affect other code),
but the command line parameter can override those defaults in
update-codegen.sh.

Based on PR #116299 

#### Does this PR introduce a user-facing change?
```release-note
api: resource.k8s.io/v1alpha1.PodScheduling was renamed to resource.k8s.io/v1alpha2.PodSchedlingHints. This type is used internally, so no action is required.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3063
```

### CODE REVIEW

**Comment by pohly** [comment]
I have rebased this on top of https://github.com/kubernetes/kubernetes/pull/116458. The approach is now that v1alpha1.PodScheduling still exists. It can be used to clean up.

All Kubernetes components and thus also the DRA drivers use v1alpha2.PodSchedulingHints. Version skew between core Kubernetes and DRA drivers is not supported.

File: pkg/apis/resource/types.go [hunk 1]
```diff
@@ -222,6 +222,57 @@ type PodSchedulingStatus struct {
...
 
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PodSchedulingHints objects hold information that is needed to schedule
+// a Pod with ResourceClaims that use "WaitForFirstConsumer" allocation
+// mode.
+//
+// This is an alpha type and requires enabling the DynamicResourceAllocation
+// feature gate.
+type PodSchedulingHints struct {
```
Comment (thockin) [inline pkg/apis/resource/types.go:233]: "Hint" implies "optional" - is that really the case here?

Comment (pohly) [reply pkg/apis/resource/types.go:233]: It's more "hint" in the sense of "potentially incomplete information" than "feel free to ignore this".

Some fields are required. Without `spec.selectedNode`, delayed allocation will not start.

See https://github.com/kubernetes/kubernetes/issues/114283 for some other ideas. We can continue brainstorming...

- `PodSchedulingInfo`
- `PodSchedulingData`

In both cases, we probably want to use the same word for plural and singular?

/cc @sftim

**Comment by pohly** [comment]
My last rebase broke "PodSchedulingHints" - let's clarify the name before I fix it.

**Pull request closed without merging**